src/Features/CSharp/Portable/BraceCompletion/BracketBraceCompletionService.cs

```diff
@@ -0,0 +1,37 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.BraceCompletion;
+using Microsoft.CodeAnalysis.Host.Mef;
+
+namespace Microsoft.CodeAnalysis.CSharp.BraceCompletion
+{
+    [Export(LanguageNames.CSharp, typeof(IBraceCompletionService)), Shared]
+    internal class BracketBraceCompletionService : AbstractBraceCompletionService
+    {
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public BracketBraceCompletionService()
+        {
+        }
+
+        protected override char OpeningBrace => Bracket.OpenCharacter;
+
+        protected override char ClosingBrace => Bracket.CloseCharacter;
+
+        public override async Task<bool> AllowOverTypeAsync(BraceCompletionContext context, CancellationToken cancellationToken)
+        {
+            return await CheckCurrentPositionAsync(context.Document, context.CaretLocation, cancellationToken).ConfigureAwait(false)
+                && await CheckClosingTokenKindAsync(context.Document, context.ClosingPoint, cancellationToken).ConfigureAwait(false);
```

Comment:
so this logic isn't apperant to me.  why is overtype ok if the currentposition/closingtoken is ok?

for example, if the above code were just `CheckCurrentPositionAsync` or `CheckClosingTokenKindAsync` i would have no idea if that were right/wrong.

Reply:
I'm resolving this by moving these checks to a method (with much nicer name) in the base and doc'ing that method.
I also unified a bunch of the AllowOverType implementations which should make everything clearer (and resolve the other allowovertypeasync comments)

Reply:
excellent!